### ISSUE TITLE
[ADD] environment checks: Where add environment checks?

### ISSUE THREAD

**Issue by moylop260** [OCA/maintainer-quality-tools#331]
# Previous info

We have a [pylint-odoo](https://github.com/oca/pylint-odoo) tools to check common fails or guidelines in odoo modules, just in the context of the project.

We want add other environment checks, It where we need a full environment to check them.

By example:
- [ ] `import-error` We can't import a library python added in a sentence `import mylibrary`.
- [ ] `xml-id-unachievable` The dependencies tree don't is unachievable a xml_id used.
- [ ]  `incoherent-ref-xml-id` A xml_id of demo is used in a xml of data.
- [ ] `inherit-method-wo-super` A method inherited from other class or module and the `super` is not used.

An so on.

This checks is not detectable by odoo or is too slow to detect them.
The build `UNIT_TESTS=1` should detect the second one but is too too slow.
# Solutions
1. We can add this checks in pylint-odo and add it in the `lint` build of `travis` the full installation of clone oca dependencies and pip install requirement.txt.
2. Or we can add this checks in a new module of odoo and add it in the `tests` build of `travis` a installation of this module. (We have the full environment in this case).
3. Other one... (your proposal)
# Advantages and disadvantages

| Solution | Advantages | disadvantages |
| --- | --- | --- |
| pylint-odoo | Faster and practical, we don't need a database to check them<br/> | Hard to develop<br/>Download and install all repositories and `requirements.txt` |
| module of odoo | Easy to develop<br/>We have all repositories and requirements.txt | Slow and impractical, we need a database. |

Before starting we need to know the correct way and avoid a -1 or a `discourage` comment after developed.

Thanks in advance.

Sincerely I'm agree with both solutions 😞 and I can't decide it.

**Comment by moylop260** [comment]
@lepistone @guewen @pedrobaeza @dreispt @max3903 @dufresnedavid @cyrilgdn @lmignon
Your feedback is welcome

**Comment by lepistone** [comment]
As usual, thanks for your efforts @moylop260 !

Generally speaking, pylint performs static analysis, while the "odoo module" approach is more like introspection of a running system.

For static languages like Java, the compiler and the IDE can understand many things about the system at compile time, so static analysis can do many checks reliably.

On the other hand, with python everything is dynamic by design, and code could do really everything at runtime, including importing files, changing classes and so on. The ORM for example does a lot of this magic internally, which is nice.

So even though tools like pylint, flake8 and pycharm can do some analysis, only some can be done reasonably, and we cannot be sure it will cover all cases. A missing import for example could be added automatically by a test framework like behave, or an XML id could be computed at runtime...

All that to say: I feel static analysis is not only hard, but also potentially inaccurate and not very useful.

Also, I think the first 3 points you mention make the server crash on startup or when running the code. In other words, our builds are red already in those cases. I think that is enough.

Regarding `inherit-method-wo-super`, in some cases not using `super` is what you actually want.

Thanks!

**Comment by pedrobaeza** [comment]
I also also very difficult to achieve the static analysis, so option 2 seems the best approach. About the super question, I don't agree with @lepistone: you must never break the inheritance chain, or at least not without providing a fallback (look for example https://github.com/OCA/bank-payment/pull/226/files#diff-6b9973ae40111c4bd9be31ef0991f1c3R28), so it's reasonable to have this check. If it must be required or only a warning can be discussed later, but for me it's a blocking one. You should explicitly deactivate the check or provide something like my example.

**Comment by moylop260** [comment]
Thanks for reply.

Very good comments @Lepistone, thanks.

+1 with @Pedro.
This errors that I used is because odoo don't show a red fail at runtime.

- If you use a `demo` xml_id from a data ref, odoo don't show error because
our CI is started with data demo.
But when you want use this same module --without-demo=all you will have
runtime errors and you will need a new pr to fix it, because we don't
detected them early.

- If you use a xml_id that for my module is unachievable, but installed
with other ones is achievable, you will see a error when installing with
`-i mymodule` but in the ci is installed with `-i all_other_modules` and
the other module make achievable the xml_id.

This checks is hard... but don't is so hard... if you give me your support
to develop I can start the work... but if you think that this one is not
very useful I will abort it.

What do you think?

**Comment by moylop260** [comment]
Thanks for reply.

Very good comments @Lepistone, thanks.

+1 with @pedrobaeza 
There are errors that odoo don't show a red fail at runtime.

- Examples:

 - If you use a `demo` xml_id from a data ref, odoo don't show error because our CI is started with data demo.
But when you want use this same module `--without-demo=all` you will have runtime errors and you will need a new pr to fix it, because we don't detected them early.

 - If you use a xml_id that for my module is unachievable, but installed with other ones is achievable, you will see a error when installing with `-i mymodule` but in the ci is installed with `-i all_other_modules` and the other module make achievable the xml_id.

Those checks are hard... but don't are so hard... if you give me your support to develop I can start the work... but if you think that this one is not very useful I will abort it.

What do you think?

**Comment by dreispt** [comment]
I agree with @lepistone but don't share all his conclusions.
Some of the checks you mention are useful for static analysis, such as the XML Ids.
I would love to see the day where I have it integrated with my editor and it warns me I mistyped an XML Id.
Exceptions will inevitably exist for any rule, so the key is that we should be able to add a marker disabling a specific check on a file.

**Comment by dreispt** [comment]
Both approaches are complementary and each has it's own use case:
- pylint should aid the developer to detect errors faster, avoiding additional upgrade-error-fix cycles
- odoo module is for improved consistency tests, not easily detectable otherwise, and too hard to implement by pylint

I agree with @lepistone but don't share all his conclusions.
Some of the checks you mention are useful for static analysis, such as the XML Ids.
I would love to see the day where I have it integrated with my editor and it warns me I mistyped an XML Id.
Exceptions will inevitably exist for any rule, so the key is that we should be able to add a marker disabling a specific check on a file.

**Comment by lepistone** [comment]
OK @dreispt, reading myself again I sound more negative that I wanted.

Having checks in my editor and trying to fix them all before running the code is a great part of my workflow and I'd have trouble without. So yes I'm all for static checks actually! I'm sorry I sounded too pessimistic.

One thing I'm skeptical about is trying too hard to give the checks a specific knowledge of the system by simulating it. I am thinking about YAML tests which I think failed because they were trying to simulate parts of the system (like interaction with views and onchanges), and the result was finding bugs in the simulated code, or not seeing actual bugs in the system.

One think that I'd love to do is to find a way to tell the checker to somehow run in an environment where the odoo modules have been loaded. This is because when you start Odoo it does magic in the import system so that when you `self.env['sale.order']` the right import is done and so on.

Maybe it is possible to tell the checker to construct the environment like `odoo shell` so that it could tell immediately if an xmlid or a module are not in the environment? This could maybe allow IDEs to make useful suggestions. To do that, we would need to know where the configuration file is (at least for the `addons_path`). Just a thought...

**Comment by dreispt** [comment]
@lepistone I think those "advanced checks" are easier implemented with @moylop260 's option 2.
So, it's not really an either/or choice, but rather two complementary tools.
Technical challenges will decide which feature should go where.

**Comment by moylop260** [comment]
@dreispt @lepistone 
Thanks for you valuable feedback.

I'll try create this "advanced checks" from option 2 (module of odoo) to have a faster Proof Of Concept, after we can think if is posible migrate it to `pylint-odoo`.

Thank you again!

**Comment by moylop260** [comment]
FYI the first version of the runtime lint module: https://github.com/OCA/server-tools/pull/444
Your feedback is welcome!

**Comment by nhomar** [comment]
@moylop260 Is this issue yet valid?

can you point us what's necessary to happen to close it in your side?

**Comment by moylop260** [comment]
I need a new project (called dylint-odoo) in order to create the modules to create dynamic checkers by each odoo version and this modules should be downloaded from MQT.

**Comment by pedrobaeza** [comment]
Strange abbreviation. Isn't better to call it `dyn-lint-odoo` or without abbreviation at all: `dynamic-lint-odoo`?

**Comment by moylop260** [comment]
`dynamic-lint-odoo` is good to me
